jupyterhub/handlers/base.py

```diff
@@ -80,9 +81,11 @@ async def prepare(self):
         The current user (None if not logged in) may be accessed
         via the `self.current_user` property during the handling of any request.
         """
-        self.scopes = set()
+        self.raw_scopes = set()
+        self.parsed_scopes = {}
         try:
             await self.get_current_user()
+            scopes.build_scope_schema(self)
```

Comment:
It feels a tiny bit off to me to pass the handler itself to the module utility function. My general feeling is that the structure should be:

- Handler retrieves 'requester' i.e. the APIToken if token-authenticated, or just the User itself if cookie-authenticated
- that object is passed to the utility, which *returns*
- Handler deals with assigning the results to attributes

If a standalone function takes an Instance as a single argument and performs several operations on that instance, *especially attribute assignment*, that's a good sign to me that it should be a method instead. If you think separating it like I described above would not be an improvement, keeping it as-is and moving it back to a method is AOK, but I like the do like the idea of making it standalone as you have it.

What do you think?

Reply:
I think that makes more sense indeed. I have refactored the logic a bit, now these functions are 'pure' and the handler takes care of the assignments